lib/Sema/TypeCheckDecl.cpp

```diff
@@ -5581,6 +5583,152 @@ static Optional<ObjCReason> shouldMarkClassAsObjC(TypeChecker &TC,
+  // context as well as stuff like Optionals
+  if (auto BGT = dyn_cast<BoundGenericType>(desugared)) {
+    for (auto T: BGT->getGenericArgs()) {
+      if(isTypePresentForCircularity(T, known)) {
+        return true;
+      }
+    }
+
+    // check the overall generic type
+    auto BGTT = BGT->getDecl()->getDeclaredTypeInContext();
+    if (known.find(BGTT.getPointer()) != known.end()) {
+      return true;
+    }
+
+    // check this generic type's members.
+    return checkDeclCircularityImpl(BGT->getDecl(), known);
+  }
+
+  // if this is a tuple, deconstruct it and check each element
+  if (auto tuple = dyn_cast<TupleType>(desugared)) {
+    for (auto Elt: tuple->getElements()) {
+      if (isTypePresentForCircularity(Elt.getType(), known)) {
+        return true;
+      }
+    }
+  }
+
+  if (known.find(desugared) != known.end()) {
+    return true;
+  }
+  return false;
+}
+
+
+// Given a declaration, iterate over its members and make sure non of them
+// is of a value type of the very same declaration. Use
+// 'isTypePresentForCircularity' to exam related types of each member.
+bool checkDeclCircularityImpl(Decl *decl, std::set<TypeBase *> &known) {
+
+  // skip unintesting declarations...
+
+  // ...implicit members...
+  if (decl->isImplicit()) { return false; }
+
+  // ...static memebers, non-value types...
+  if (auto s = dyn_cast<VarDecl>(decl)) {
+    if (s->isStatic() || !s->hasStorage()) { return false; }
+  }
+
+  bool isEnum = false;
+  bool isStruct = isa<StructDecl>(decl);
+
+  // ... and enums marked 'indirect'
+  if (auto e = dyn_cast<EnumDecl>(decl)) {
+    if (e->isIndirect()) {
+      return false;
+    }
+    isEnum = true;
+  }
+
+  if (auto nominal = dyn_cast<NominalTypeDecl>(decl)) {
+    auto dt = nominal->getDeclaredTypeInContext();
+    // record existence of this struct or enum, if we see it in value member
+    // types, subtypes, etc, complain.
+    if (isEnum || isStruct) {
+      known.insert(dt.getPointer());
+    }
+
+    // recurse for nested struct/enums as well as normal member declarations.
+    for (auto m: nominal->getMembers(false)) {
+      if (isa<VarDecl>(m) ||
+          isa<EnumElementDecl>(m) ||
+          isa<StructDecl>(m) ||
+          isa<EnumDecl>(m))
+        if (checkDeclCircularityImpl(m, known)) {
+          return true;
+        }
+    }
+
+    // sibling nested types are okey, don't count them
+    if (isEnum || isStruct) {
+      known.erase(dt.getPointer());
+    }
+  } else {
+    if (auto vd = dyn_cast<VarDecl>(decl)) {
+      if (vd->hasType()) {
+        auto vdt = vd->getType();
+        if (!vdt->hasReferenceSemantics() && // reference types are fine
+            isTypePresentForCircularity(vdt, known) ) {
+          return true;
+        }
+      }
+    }
+
+    if (auto ed = dyn_cast<EnumElementDecl>(decl)) {
+      // indirect cases, cases without arugment types are fine, skip these
+      if (!ed->isIndirect() && ed->hasArgumentType())
```

Comment:
Maybe move the EnumDecl->isIndirect check here, from line 5658? Then you won't need to special-case EnumDecl up there.